lib/path/default.nix

```diff
@@ -100,6 +103,54 @@ let
     # An empty string is not a valid relative path, so we need to return a `.` when we have no components
     (if components == [] then "." else concatStringsSep "/" components);
 
+  # Type: Path -> { path :: Path, root :: Path, components :: [ String ] }
+  #
+  # Deconstruct a path value type into:
+  # - path: The full path
+  # - root: The filesystem root of the path, generally `/`
+  # - components: All the path's components
+  #
+  # This is similar to `splitString "/" (toString path)` but safer
+  # because it can distinguish different filesystem roots
+  deconstructPath = path:
+    let
+      recurse = components: base:
+        # If the parent of a path is the path itself, then it's a filesystem root
+        if base == dirOf base then { root = base; inherit path components; }
+        else recurse ([ (baseNameOf base) ] ++ components) (dirOf base);
+    in recurse [] path;
+
+  # Type: String -> (DeconstructedPath -> DeconstructedPath -> a) -> (Path -> (Path -> a))
+  #
+  # Turn a binary path function on deconstructed paths into a function on
+  # non-deconstructed paths. The resulting function takes one path after
+  # another, deconstructing it, and caching it before taking the next path as
+  # an argument, allowing the deconstruction to be reused between calls with
+  # the same first argument. With both paths passed, a check ensures that they
+  # have the same filesystem root, before the passed function on deconstructed
+  # paths is called. The first argument is the context for error messages.
+  withTwoDeconstructedPaths = context: f:
+    path1:
+    assert assertMsg
+      (isPath path1)
+      "${context}: First argument is of type ${typeOf path1}, but a path was expected";
+    let
+      deconPath1 = deconstructPath path1;
+    in
+      path2:
+      assert assertMsg
+        (isPath path2)
+        "${context}: Second argument is of type ${typeOf path2}, but a path was expected";
+      let
+        deconPath2 = deconstructPath path2;
+      in
+        assert assertMsg
+        (deconPath1.root == deconPath2.root) ''
+          ${context}: Filesystem roots must be the same for both paths, but paths with different roots were given:
+              first argument: "${toString path1}" (root "${toString deconPath1.root}")
+              second argument: "${toString path2}" (root "${toString deconPath2.root}")'';
+        f deconPath1 deconPath2;
```

Comment:
This is way too clever to be maintainable.

Every day it looks more like what you actually want to implement is error tracing and type checking at the Nix language level.

Apparently what we really need here is
1. generic type assertions, this seems reasonable to extract into a helper function
2. the root check, but there is no reason to hard-code it to be about pairs of paths

   actually it's specific enough to duplicate it across the two functions, as it was originally.

Reply:
I agree.

> generic type assertions

This would perhaps be nice to have in the language, although we'd have to choose between having a static type system or slow evaluation or more infinite recursions. I won't go into more detail here.

To move this forward, I suggest removing this function, which is a non-abstraction, and use more general checks at the call sites.